### PR TITLE
alsa-ucm-conf: apply patch to fix SplitPCM: Device argument may not b…

### PR DESCRIPTION
**Pull request by Sporesirius**
## Description of changes

This patch fixes some audio devices working with alsa-ucm-conf that use SplitPCM.
Upstream fix: https://github.com/alsa-project/alsa-ucm-conf/commit/b68aa52acdd2763fedad5eec0f435fbf43e5ccc6

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Sporesirius** [comment]
Hey @IogaMaster I watched your video about NixOS and I would like to take you up on your offer of a PR review.

**Review by IogaMaster — APPROVED** [review]
Looks great! Thanks for watching my video!

**Comment by oddlama** [comment]
LGTM
I've encountered the same problem on one of my devices, and applying this patch fixes it.

**Review by oddlama — APPROVED** [review]

**Comment by Sporesirius** [comment]
Can this be merged, or is there a problem?

**Comment by icyqar** [comment]
Needed to fix issue GoXLR-on-Linux/goxlr-utility/issues/138

File: pkgs/by-name/al/alsa-ucm-conf/package.nix
Comment (roastiek) [inline pkgs/by-name/al/alsa-ucm-conf/package.nix:17]: Could you add a comment, that the patch should be removed by the next package upgrade as it is from the master branch?

**Review by roastiek — COMMENTED** [review]
Looks ok.

File: pkgs/by-name/al/alsa-ucm-conf/package.nix
Comment (Sporesirius) [reply pkgs/by-name/al/alsa-ucm-conf/package.nix:19]: Done

**Review by oddlama — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by roastiek — APPROVED** [review]

**Comment by errnoh** [comment]
Any plans to backport this to 23.11 once it's merged? This issue causes some audio interfaces that were working in 23.05 to not work properly in 23.11.

**Comment by Sporesirius** [comment]
@errnoh I would love that. AFAIK from the CONTRIBUTING.md only a nixpkgs maintainer can backport. But what I'm wondering is why the "approved by: package-maintainer" label is not set, since I thought @roastiek was the maintainer of alsa-ucm-conf?
I think we have to wait for a nixpkgs committer anyway.

**Comment by Sporesirius** [comment]
Hi @SuperSandro2000, can you please review this PR.

**Comment by oddlama** [comment]
This PR will cause a whole lot of rebuilds because it's a dependency of alsa-lib, so maybe this should target staging and not master.

**Comment by Sporesirius** [comment]
@oddlama I tried to follow CONTRIBUTING.md to rebase to staging. Can you check if this is okay?

**Comment by SuperSandro2000** [comment]
This should really be one comment otherwise LGTM

**Review by oddlama — APPROVED** [review]

**Review by IogaMaster — APPROVED** [review]

**Comment by Sporesirius** [comment]
@SuperSandro2000 I squashed the commits. Hope it's good now.

**Review by IogaMaster — APPROVED** [review]

**Comment by arh78750** [comment]
please forgive my nix noob question. This fixes and issue I currently have with my goxlr. I was able to get this patch and apply it to the nixpkgs in my nix config but because so much depends on alsa it tried to rebuild like half the packages on my system from source. Is there any way for me to somehow apply to patch to the package but let all the other things on the system build with the old version? https://github.com/arh78750/nixosconfig/blob/d3d6f1ad90dfe5d626f9309a927cb93534af7a80/flake.nix#L19 here is where I'm applying the patch.

**Comment by oddlama** [comment]
@arh78750 Since this is a very simple one-byte patch of a configuration file, I personally just manually edited the file of the broken package in the nix-store by temporarily mounting it rw. In practice you should never ever do that, but in this specific case it's an acceptable workaround because it doesn't affect dependent packages and would otherwise cause too many rebuilds. Just be prepared to redo it when you update your system.

**Comment by roastiek** [comment]
You only need update settings of pipewire or pulseaudio service. Other applications don't read ucm configs directly.

You can change theirs `package` option with recompiled original againts patches alsa/alsa-ucm-conf.

Or setup `ALSA_CONFIG_UCM2` env variable in those services to the new patched config. I think that it worked at leats with pulseaudio two years ago.

**Comment by SuperSandro2000** [comment]
Seems unrelated to me. Probably something else was broken on staging.

@ofborg build alsa-ucm-conf
@ofborg eval

**Comment by Sporesirius** [comment]
@ofborg build alsa-ucm-conf
@ofborg eval

**Comment by Sporesirius** [comment]
Hey @roastiek, can you review, pls.
Is there anything that needs to be taken care of or can it be merged?

**Review by roastiek — APPROVED** [review]

**Review by ghost — APPROVED** [review]
https://github.com/alsa-project/alsa-ucm-conf/issues/346
LGTM

**Comment by Sporesirius** [comment]
@a-n-n-a-l-e-e thanks for the merge. Is it possible to backport this to 23.11?

**Pull request merged**
